client.go

```diff
@@ -1,219 +1,101 @@
 package apns
 
 import (
-	"container/list"
 	"crypto/tls"
-	"io"
-	"log"
+	"sync"
 	"time"
 )
 
-type buffer struct {
-	size int
-	*list.List
-}
-
-func newBuffer(size int) *buffer {
-	return &buffer{size, list.New()}
-}
-
-func (b *buffer) Add(v interface{}) *list.Element {
-	e := b.PushBack(v)
-
-	if b.Len() > b.size {
-		b.Remove(b.Front())
-	}
-
-	return e
-}
-
+// Client creates a session with Apple and handles reconnection.
 type Client struct {
-	Conn         *Conn
-	FailedNotifs chan NotificationResult
+	conn Conn
 
-	notifs chan Notification
-	id     uint32
+	sess  Session
+	sessm sync.Mutex
 }
 
-func newClientWithConn(gw string, conn Conn) Client {
-	c := Client{
-		Conn:         &conn,
-		FailedNotifs: make(chan NotificationResult),
-		id:           uint32(1),
-		notifs:       make(chan Notification),
-	}
+func newClientWithConn(conn Conn) (Client, error) {
+	c := Client{conn: conn}
 
-	go c.runLoop()
+	sess := newSession(conn)
+	err := sess.Connect()
+	if err != nil {
+		return c, err
+	}
 
-	return c
+	return Client{conn, sess, sync.Mutex{}}, nil
 }
 
-func NewClientWithCert(gw string, cert tls.Certificate) Client {
+// NewClientWithCert creates a client of the Apple gateway given a certificate.
+func NewClientWithCert(gw string, cert tls.Certificate) (Client, error) {
 	conn := NewConnWithCert(gw, cert)
-
-	return newClientWithConn(gw, conn)
+	return newClientWithConn(conn)
 }
 
+// NewClient is a helper that accepts a certificate/key pair.
 func NewClient(gw string, cert string, key string) (Client, error) {
 	conn, err := NewConn(gw, cert, key)
 	if err != nil {
 		return Client{}, err
 	}
 
-	return newClientWithConn(gw, conn), nil
+	return newClientWithConn(conn)
 }
 
+// NewClientWithFiles is a helper that loads a certificate/key from files.
 func NewClientWithFiles(gw string, certFile string, keyFile string) (Client, error) {
 	conn, err := NewConnWithFiles(gw, certFile, keyFile)
 	if err != nil {
 		return Client{}, err
 	}
 
-	return newClientWithConn(gw, conn), nil
+	return newClientWithConn(conn)
 }
 
+// Send a notification, handling disconnections.
 func (c *Client) Send(n Notification) error {
-	c.notifs <- n
-	return nil
-}
-
-func (c *Client) reportFailedPush(v interface{}, err *Error) {
-	failedNotif, ok := v.(Notification)
-	if !ok || v == nil {
-		return
+	if c.sess.Disconnected() {
+		c.reconnectAndRequeue()
```

Comment:
I'm not sure if I like waiting for the next Send before requeueing - at least in the case where sends are more sporadic.

On master these are sent in the background after the failure.